modules/recon/hosts-ports/shodan_ip.py

```diff
@@ -1,32 +1,36 @@
 from recon.core.module import BaseModule
-from recon.mixins.search import ShodanAPIMixin
-import re
+import shodan
 
 
-class Module(BaseModule, ShodanAPIMixin):
+class Module(BaseModule):
 
     meta = {
         'name': 'Shodan IP Enumerator',
         'author': 'Tim Tomes (@lanmaster53) and Matt Puckett (@t3lc0)',
-        'version': '1.0',
-        'description': 'Harvests port information from the Shodan API by using the \'ip\' search operator. Updates the \'ports\' table with the results.',
+        'version': '1.1',
+        'description': 'Harvests port information from the Shodan API by using the \'ip\' search operator. Updates the '
+                       '\'ports\' table with the results.',
         'required_keys': ['shodan_api'],
         'query': 'SELECT DISTINCT ip_address FROM hosts WHERE ip_address IS NOT NULL',
         'options': (
             ('limit', 1, True, 'limit number of api requests per input source (0 = unlimited)'),
         ),
+        'dependencies': ['shodan']
     }
 
     def module_run(self, ipaddrs):
         limit = self.options['limit']
+        api = shodan.Shodan(self.keys.get('shodan_api'))
         for ipaddr in ipaddrs:
             self.heading(ipaddr, level=0)
-            query = f"ip:{ipaddr}"
-            results = self.search_shodan_api(query, limit)
-            for host in results:
-                address = host['ip_str']
-                port = host['port']
-                if not host['hostnames']:
-                    host['hostnames'] = [None]
-                for hostname in host['hostnames']:
-                    self.insert_ports(ip_address=address, port=port, host=hostname)
+            try:
+                ipinfo = api.host(ipaddr)
+                for port in ipinfo['data']:
+                    try:
+                        for hostname in port['hostnames']:
+                            self.insert_ports(host=hostname, ip_address=ipaddr, port=port['port'],
+                                              protocol=port['transport'])
+                    except KeyError:
+                        self.insert_ports(ip_address=ipaddr, port=port['port'], protocol=port['transport'])
+            except shodan.exception.APIError:
```

Comment:
Should we do some level of reporting here? What causes this type of error?

Reply:
The shodan module raises this error if an IP is not found in its database so we either catch it like this which how the rest of the modules seem to handle no data or we print a message saying not found in DB.

Reply:
OK. Cool. I like this approach then. Thanks!